sqlglot/optimizer/remove_unused_columns.py

```diff
@@ -0,0 +1,162 @@
+    Search SELECT for columns that can be removed.
+
+    Returns:
+         Same as `_pushdown_statement`
+    """
+    scope = select_scope(expression)
+
+    # Collect a map of all referenced columns
+    columns = {}
+    for column in scope.columns:
+        selectable_name = column.text("table")
+        column_name = column.text("this")
+        if not selectable_name:
+            msg = (
+                "Expected all columns to have table prefixes. "
+                "Did you run 'qualify_columns' first?\n"
+                f"Received: {column_name}"
+            )
+            raise OptimizeError(msg)
+
+        # Use the Expression identity for the key since Expressions are hashed by value.
+        columns[id(column)] = column
+
+    # Collect all the selections
+    if not expression.args.get("distinct"):
+        columns = _remove_unused_columns(expression, columns, parent_selections)
+
+    for subquery in scope.subqueries:
+        # Subqueries (as opposed to "derived_tables") aren't "selectable".
+        # So the none of the columns in the current scope to reference these.
+        _pushdown_statement(subquery, SELECT_ALL)
+
+    # Now that we've removed all the unused columns from the selections, let's
+    # build a map of all the columns we're selecting from derived tables.
+    derived_table_selections = defaultdict(set)
+    for column in columns.values():
+        derived_table_selections[column.text("table")].add(column.text("this"))
+
+    for subquery in scope.derived_tables:
+        _pushdown_statement(subquery.this, derived_table_selections[subquery.alias])
+
+    _pushdown_ctes(scope.ctes, derived_table_selections)
+
+    # Push the selections back UP so they can be used by CTEs in outer queries
+    return derived_table_selections
+
+
+def _pushdown_ctes(ctes, selections):
+    if not ctes:
+        return
+
+    # Iterate in reversed order as a CTE can reference outputs in previous CTEs
+    for cte in reversed(ctes):
+        selections = _merge_selections(
+            selections, _pushdown_statement(cte.this, selections.get(cte.alias, set()))
+        )
+
+
+def _remove_unused_columns(expression, columns, parent_selections):
+    columns = copy(columns)
+
+    new_selections = []
+    for selection in expression.selects:
+        if not isinstance(selection, exp.Alias):
+            msg = (
+                "Expected all selections to have aliases. "
+                "Did you run 'qualify_columns' first?\n"
+                f"Received: {selection}"
+            )
+            raise OptimizeError(msg)
+
+        if parent_selections == SELECT_ALL or selection.alias in parent_selections:
+            new_selections.append(selection)
+        else:
+            # Pop the column out of the set of all columns.
+            # Later, we'll use this set of columns to pushdown the selected columns from inner queries.
+            for column in selection.find_all(exp.Column):
+                columns.pop(id(column))
```

Comment:
why do id? why not just put column into the set as is?

Reply:
```python
hash(exp.Column(this="a")) == hash(exp.Column(this="a"))
```
We need to be able to remove the exact column instance from this set. So I'm using `id` instead.